### PR TITLE
[Config] allow changing the path separator

### PR DESCRIPTION
**Pull request by bburnichon**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #22268
| License       | MIT
| Doc PR        | 

In bundles configuration, no dots are allowed in key names, but as a standalone config, there is no such limitation that should be enforced.

By using current `NodeBuilder` it should be possible to change path separator used (for example: `/`)

### CODE REVIEW

File: src/Symfony/Component/Config/Definition/BaseNode.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Definition/BaseNode.php]: since all other props are protected, should we make this one protected also, and use it directly in child nodes?

File: src/Symfony/Component/Config/Definition/Builder/TreeBuilder.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Definition/Builder/TreeBuilder.php]: this should be deprecated instead with an annotation (no need for runtime BC layer since this is useless, WDYT?)

File: src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php]: protected?

File: src/Symfony/Component/Config/Definition/BaseNode.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Definition/BaseNode.php]: this should be kept, but the dot replaced by the pathSeparator

File: src/Symfony/Component/Config/Definition/Builder/TreeBuilder.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Definition/Builder/TreeBuilder.php]: I think doing removing this an doing the check inline would be fine

**Review by nicolas-grekas — COMMENTED** [review]
missing tests and CHANGELOG entry

File: src/Symfony/Component/Config/Definition/BaseNode.php
Comment (bburnichon) [inline src/Symfony/Component/Config/Definition/BaseNode.php]: Why not, but this is the only place where the name is checked. In prototyped node, name property is not enforced and make Node inconsistent.

File: src/Symfony/Component/Config/Definition/Builder/TreeBuilder.php
Comment (bburnichon) [inline src/Symfony/Component/Config/Definition/Builder/TreeBuilder.php]: It is currently called twice. It is done this way to keep code DRY.

File: src/Symfony/Component/Config/Definition/BaseNode.php
Comment (stof) [inline src/Symfony/Component/Config/Definition/BaseNode.php]: I would keep it private if it does not need to be protected. It is easier for our BC policy

File: src/Symfony/Component/Config/Definition/BaseNode.php
Comment (nicolas-grekas) [reply src/Symfony/Component/Config/Definition/BaseNode.php]: @bburnichon would you mind reverting this change (and using the $pathSeparator of course). Better not remove an existing check.

**Comment by nicolas-grekas** [comment]
Rebase needed (we don't merge PRs when they contain merge commit)

**Comment by bburnichon** [comment]
Tests are failing on Form Component which I did not change. Is it a known issue?

**Review by nicolas-grekas — APPROVED** [review]
(tests back to green)

File: src/Symfony/Component/Config/Definition/BaseNode.php
Comment (xabbuh) [inline src/Symfony/Component/Config/Definition/BaseNode.php]: Didn't we recently remove all comments like this?

File: src/Symfony/Component/Config/Definition/Builder/ChildrenAwareNodeDefinitionInterface.php
Comment (xabbuh) [inline src/Symfony/Component/Config/Definition/Builder/ChildrenAwareNodeDefinitionInterface.php]: This reads weird to me, what about `getChildNodeDefinitions()` or `getChildNodesDefinitions()`?

Comment (xabbuh) [reply src/Symfony/Component/Config/Definition/Builder/ChildrenAwareNodeDefinitionInterface.php]: I wonder if we really need this interface or if we shouldn't rather check for the method to exist. We could then add this method to the `ParentNodeDefinitionInterface` in 4.0.

File: src/Symfony/Component/Config/Definition/BaseNode.php
Comment (bburnichon) [reply src/Symfony/Component/Config/Definition/BaseNode.php]: Maybe it was reintroduced by my rebase. I will remove it.

File: src/Symfony/Component/Config/Definition/Builder/ChildrenAwareNodeDefinitionInterface.php
Comment (bburnichon) [reply src/Symfony/Component/Config/Definition/Builder/ChildrenAwareNodeDefinitionInterface.php]: I don't have strong opinion about how to name this method. Actually, I already had a hard time finding this one. The same is true for the whole interface.
In case checking for the method to exist, how would you mark deprecation notice though?

File: src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php]: `if ($this instanceof ArrayNodeDefinition) {` would be enough for the prupose, isn't it?
about the method, my preference is `getChildNodeDefinitions`

**Comment by xabbuh** [comment]
We will also need to update the `UPGRADE-3.4.md` (explain that implementing the interface without the added method is deprecated) and `UPGRADE-4.0.md` (mention the added method) files.

File: UPGRADE-4.0.md
Comment (xabbuh) [inline UPGRADE-4.0.md]: I would rewrite this to be like this:

> Added the `getChildNodeDefinitions()` method to `ParentNodeDefinitionInterface`.

**Review by xabbuh — APPROVED** [review]

**Comment by bburnichon** [comment]
Twig and VarDumper failure don't seem to be related to my PR. Are failures expected?

**Comment by xabbuh** [comment]
The Twig failure should be fixable by rebasing on the latest `3.4` branch and the VarDumper tests are currently failing on the `3.4` branch too.

**Comment by nicolas-grekas** [comment]
Moving to 4.1 as 3.4 is now in beta, thus closed for new feats.

**Comment by bburnichon** [comment]
Failures do not seem linked to my changes.

**Comment by bburnichon** [comment]
Rebased after merge of deprecation and removal of `TreeBuilder::$builder`

**Comment by nicolas-grekas** [comment]
@bburnichon please rebase: we don't merge PR with merge commits.

**Comment by bburnichon** [comment]
Sorry I did update via github interface.

**Comment by javiereguiluz** [comment]
I like this feature, but I wonder if *"path separator"* is the best name for this option. Path is strongly related to "file paths". Maybe "name separator" or "key separator" or "children separator" etc.

**Comment by nicolas-grekas** [comment]
@javiereguiluz path is the correct term, already used a lot in the component:
```
$ grep path src/Symfony/Component/Config/ -ri | wc -l
219
```

**Review by javiereguiluz — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @bburnichon.

**Pull request merged**
